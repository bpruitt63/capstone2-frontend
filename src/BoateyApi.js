import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL;


class BoateyApi {

    static token;

    /** Sends request to database and catches any errors */
    static async request(endpoint, data={}, method="get") {
        const url = `${BASE_URL}/${endpoint}`;
        const headers = {Authorization: `Bearer ${BoateyApi.token}`};
        const params = (method === "get") ? data : {};

        try {
            return (await axios({url, method, data, params, headers})).data;
        } catch (err) {
            console.error("API Error:", err.response);
            let message;
            !err.response ? message = "Server error, please try again later"
                : message = err.response.data.error.message;
            throw Array.isArray(message) ? message : [message];
        };
    };

    /** Register new user and get JWT */
    static async register(data) {
        const res = await this.request('users/register', data, 'post');
        return res.token;
    };

    /** Log in existing user and get JWT. 
     * Returns user location and preferred unit of measurement.
     */
    static async login(data) {
        const res = await this.request('users/login', data, 'post');
        const token = res.token;
        const location = res.location;
        const units = res.units;
        return {token, location, units};
    };

    /** Gets all info on a single user */
    static async getUser(username) {
        const res = await this.request(`users/${username}`);
        return res.user;
    };

    /** Updates selected data on a single user and returns all data */
    static async updateUser(username, data) {
        const res = await this.request(`users/${username}`, data, 'patch');
        return res.user;
    };

    /** Adds location into database if not already added */
    static async addLocation(data) {
        const res = await this.request(`locations/new_location`, data, 'post');
        return res.location;
    };

    /** Adds location rating into database */
    static async addRating(username, data) {
        const res = await this.request(`locations/rating/${username}`, data, 'patch');
        return res.rating;
    };

    /** Gets user and overall ratings for a point */
    static async getRatings(username, id) {
        const res = await this.request(`locations/${id}/${username}`);
        const ratings = {overallRating: res.overallRating, 
                        userRating: res.userRating};
        return ratings;
    };

    /** Adds trip and its locations into database */
    static async addTrip(username, data) {
        const res = await this.request(`trips/${username}/new_trip`, data, 'post');
        const trip = {trip: res.trip, tripLocations: res.tripLocations};
        return trip;
    };

    /** Gets logged in user's trips from database */
    static async getUserTrips(username) {
        const res = await this.request(`trips/${username}/trips`);
        return res;
    };

    /** Gets 20 most recent trips from database */
    static async getRecentTrips() {
        const res = await this.request('trips/recent_trips');
        return res;
    };

    /** Deletes trip from database */
    static async deleteTrip(username, id) {
        const res = await this.request(`trips/${username}/${id}/delete`, {}, 'delete');
        return res;
    };

    /** Adds trip rating to database */
    static async rateTrip(username, id, data) {
        const res = await this.request(`trips/${username}/${id}/rate`, data, 'patch');
        return res;
    };
};

BoateyApi.token = localStorage.getItem("token");

export default BoateyApi;