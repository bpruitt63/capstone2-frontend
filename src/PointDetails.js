import React, {useState} from 'react';
import {Link} from 'react-router-dom';
import PointMap from './PointMap';

function PointDetails({point}) {

    const [showMap, setShowMap] = useState(false);

    const toggle = () => {
        setShowMap(!showMap);
    };

    return (
        <>
            {point.images.data[0] && 
                <img src={point.images.data[0].thumbnail_url} 
                        alt={point.name} />}
            <ul>
                <span>Fuel available:</span>
                <li>Diesel?  {point.fuel.has_diesel ? 'Yes' : 'No'}</li>
                <li>Gas?  {point.fuel.has_gas ? 'Yes' : 'No'}</li>
                <li>Propane?  {point.fuel.has_propane ? 'Yes' : 'No'}</li>
            </ul>
            {point.rating && 
                <p>Marinas.com rating: {point.rating} out of {point.review_count} reviews</p>}
            <Link to={{pathname: `${point.web_url}`}} target='blank'>
                    View on Marinas.com
            </Link>
            <button onClick={toggle}>
                {showMap ? 'Hide Map' : 'Show Map'}
            </button>
            {showMap && <PointMap longLat={[point.location.lon, point.location.lat]}
                                    icon_url={point.icon_url} />}
        </>
    );
};

export default PointDetails;