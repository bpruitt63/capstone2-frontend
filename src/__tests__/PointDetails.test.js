import React from 'react';
import { render, screen } from '@testing-library/react';
import PointDetails from '../PointDetails';
import BoateyApi from '../BoateyApi';
import { BrowserRouter } from 'react-router-dom';

jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
    Map: () => ({}),
}));

jest.mock('../BoateyApi');

const point = {id: 'test1', 
                kind: 'marina',
                name: 'Test Location',
                review_count: 0,
                web_url: 'https://google.com',
                location: {lat: 40, lon: 10},
                fuel: {has_diesel: false,
                        has_propane: false,
                        has_gas: true},
                images: {data: []}};

test('renders', () => {
    render(<BrowserRouter><PointDetails point={point}/></BrowserRouter>);
});

test('matches snapshot', () => {
    const {asFragment} = render(<BrowserRouter><PointDetails point={point}/></BrowserRouter>);
    expect(asFragment()).toMatchSnapshot();
});

test('renders recent trip', () => {
    render(<BrowserRouter><PointDetails point={point}/></BrowserRouter>);
    expect(screen.getByText(/Propane:/)).toBeInTheDocument;
    expect(screen.queryByText(/Rating:/)).not.toBeInTheDocument;
});

test('renders user trip', async () => {
    BoateyApi.getRatings.mockResolvedValue({overallRating: '4.50', userRating: '#'});
    render(<BrowserRouter><PointDetails point={point} username={'test'}/></BrowserRouter>);
    await screen.findByText('4.50');
    expect(screen.getByText(/Propane:/)).toBeInTheDocument;
    expect(screen.getByText(/4.50/)).toBeInTheDocument;
    expect(screen.getByText(/#/)).toBeInTheDocument;
    expect(screen.getByText(/Submit Rating/)).toBeInTheDocument;
});