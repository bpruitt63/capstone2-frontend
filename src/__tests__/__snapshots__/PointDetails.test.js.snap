// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <ul>
    <span>
      Fuel available
    </span>
    <li>
      Diesel:  No
    </li>
    <li>
      Gas:  Yes
    </li>
    <li>
      Propane:  No
    </li>
  </ul>
  <br />
  <a
    class="link"
    href="https://google.com"
    target="blank"
  >
    View on Marinas.com
  </a>
  <button
    class="pointButton btn btn-secondary"
    type="button"
  >
    Show Map
  </button>
  <br />
</DocumentFragment>
`;
